k2/python/csrc/torch/fsa_algo.cu

```diff
@@ -686,72 +699,93 @@ static void PybindReplaceFsa(py::module &m) {
 }
 
 static void PybindCtcGraph(py::module &m) {
+  m.def(
+      "ctc_graph",
+      [](RaggedAny &symbols, torch::optional<torch::Device> = {},
+         bool modified = false,
+         bool need_arc_map =
+             true) -> std::pair<FsaVec, torch::optional<torch::Tensor>> {
+        DeviceGuard guard(symbols.any.Context());
+        Array1<int32_t> arc_map;
+        FsaVec graph = CtcGraphs(symbols.any.Specialize<int32_t>(), modified,
+                                 need_arc_map ? &arc_map : nullptr);
+        torch::optional<torch::Tensor> tensor;
+        if (need_arc_map) tensor = ToTorch(arc_map);
+        return std::make_pair(graph, tensor);
+      },
+      py::arg("symbols"), py::arg("device") = py::none(),
+      py::arg("modified") = false, py::arg("need_arc_map") = true);
+
   m.def(
       "ctc_graph",
       [](const std::vector<std::vector<int32_t>> &symbols,
-          int32_t gpu_id = -1, bool modified = false, bool need_arc_map = true)
-        -> std::pair<FsaVec, torch::optional<torch::Tensor>> {
-        ContextPtr context;
-        if (gpu_id < 0)
-          context = GetCpuContext();
-        else
-          context = GetCudaContext(gpu_id);
+         torch::optional<torch::Device> device = {}, bool modified = false,
```

Comment:
What's the benefit of passing argument device by `torch::Device` and `std::string`,  why not use `py::object` to avoid implementing it twice.

we can get context like the code below, whether the passing in argument is `cuda:0` or `torch.device`.
```
ContextPtr GetContext(py::object device_obj) {
  auto device = torch::Device(static_cast<py::str>(device_obj));
  if (device.type() == torch::kCPU) {
    return GetCpuContext();
  } else if (device.type() == torch::kCUDA) {
    return GetCudaContext(device.index());
  } else {
    K2_LOG(FATAL) << "Unsupported device: " << device.str();
    return GetCpuContext();   // unreachable code
  }
}
```

Reply:
Yes, your approach would work, but it produces less beautiful documentation.

The current implementation produces the following doc

```python
import _k2
help(_k2.linear_fsa)
```

```
linear_fsa(...) method of builtins.PyCapsule instance
    linear_fsa(*args, **kwargs)
    Overloaded function.

    1. linear_fsa(labels: k2::RaggedAny, device: Optional[torch::Device] = None) -> k2::Ragged<k2::Arc>

    2. linear_fsa(labels: List[int], device: Optional[torch::Device] = None) -> k2::Ragged<k2::Arc>

    3. linear_fsa(labels: List[int], device: Optional[str] = None) -> k2::Ragged<k2::Arc>

    4. linear_fsa(labels: List[List[int]], device: Optional[torch::Device] = None) -> k2::Ragged<k2::Arc>

    5. linear_fsa(labels: List[List[int]], device: Optional[str] = None) -> k2::Ragged<k2::Arc>
 ```

Reply:
Cool！Thanks for explanation！